File: C++/MdsOmTransports.cpp
#include "MdsOm.h"
#include "MdsOmInternal.h"

namespace MdsOmNs {

// ---------------------------------------------------------
MdsOmTransportSet::MdsOmTransportSet()
{
	listIndex = 0;
}

MdsOmTransportSet::~MdsOmTransportSet()
{
}

// ---------------------------------------------------------
MdsOmTransports::MdsOmTransports()
{
	listIndex = 0;
	env = NULL;
}

void MdsOmTransports::close()
{
	try {
		mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::close: %s delete transports map count=%d", env->getName(), transportMap.size());

		MdsOmTransportCollection::const_iterator itx;
		MdsOmTransportMap::const_iterator it;
		for (it = transportMap.begin(); it != transportMap.end(); it++) {
			MdsOmTransportSet* s = it->second;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::close: %s delete transport name=%s", env->getName(), t->getName());
				delete t;
			}
			delete s;
		}
		transportMap.clear();

		mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::close: %s delete transports list count=%d", env->getName(), transportList.size());
		for (itx = transportList.begin(); itx != transportList.end(); itx++) {
			MdsOmTransport* t = *itx;
			mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::close: %s delete transport name=%s", env->getName(), t->getName());
			delete t;
		}
		transportList.clear();

		mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::close: %s", env->getName());
	} catch (MamaStatus& status) {
		mama_log(MAMA_LOG_LEVEL_ERROR, "MdsOmTransports::close: error %s %s", env->getName(), status.toString());
		// Send original exception reference back to caller
		throw;
	}
}

MdsOmTransports::~MdsOmTransports()
{
	//
}

void MdsOmTransports::addTransport(MamaTransport* transport)
{
	// Add to list
	transportList.push_back(new MdsOmTransport(transport));
}

void MdsOmTransports::addTransport(MamaTransport* transport, string source)
{
	// Add to map
	MdsOmTransportSet* s = transportMap[source];
	if (s == NULL) {
		s = new MdsOmTransportSet();
		transportMap[source] = s;
	}
	s->transportList.push_back(new MdsOmTransport(transport));
}

void MdsOmTransports::setStatus(MamaTransport* transport, MdsOmTransportStatus status)
{
	if (isMap()) {
		MdsOmTransportMap::const_iterator it;
		for (it = transportMap.begin(); it != transportMap.end(); it++) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				if (t->getTransport() == transport) {
					t->setStatus(status);
					return;
				}
			}
		}
	} else {
		MdsOmTransportCollection::iterator itx = transportList.begin();
		while (itx != transportList.end()) {
			MdsOmTransport* t = *itx++;
			if (t->getTransport() == transport) {
				t->setStatus(status);
				return;
			}
		}
	}
	mama_log(MAMA_LOG_LEVEL_ERROR, "MdsOmTransports::setStatus: did not find transport %s %p", transport->getName(), transport);
	printTransports();
}

void MdsOmTransports::printTransports()
{
	mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::printTransports: %s map=%d list=%d", env->getName(), transportMap.size(), transportList.size());
		
	if (isMap()) {
		MdsOmTransportMap::const_iterator it;
		for (it = transportMap.begin(); it != transportMap.end(); it++) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::printTransports: %s %p", t->getTransport()->getName(), t->getTransport());
			}
		}
	} else {
		MdsOmTransportCollection::iterator itx = transportList.begin();
		while (itx != transportList.end()) {
			MdsOmTransport* t = *itx++;
			mama_log(MAMA_LOG_LEVEL_NORMAL, "MdsOmTransports::printTransports: %s %p", t->getTransport()->getName(), t->getTransport());
		}
	}
}

// Get a transport for a specific source
MamaTransport* MdsOmTransports::getTransport(string source)
{
	if (transportMap.size() == 0 && transportList.size() == 0) {
		mama_log(MAMA_LOG_LEVEL_ERROR, "MdsOmTransports::getTransport(source): error %s both collections are empty", env->getName());
		throw MdsOmStatus(MDS_OM_STATUS_NO_TRANSPORTS_AVAILABLE);
	}

	if (transportList.size() > 0) {
		return getTransport();
	} else {
		MdsOmTransportMap::const_iterator it = transportMap.find(source);
		if (it != transportMap.end()) {
			// Get a transport from the mapped round-robin list
			MdsOmTransportSet* s = it->second;
			if (s->transportList.size() == 0) {
				// Should not be empty ...
				mama_log(MAMA_LOG_LEVEL_ERROR, "MdsOmTransports::getTransport(source): error %s list is empty for map source %s", env->getName(), source.c_str());
				return NULL;
			}
			MdsOmTransport* t = s->transportList.at(s->listIndex);
			if (++s->listIndex >= s->transportList.size()) {
				s->listIndex = 0;
			}
			return t->getTransport();
		} else {
			// Source not found, return a transport from the first list
			MdsOmTransportSet* s = transportMap.begin()->second;
			MdsOmTransport* t = s->transportList.at(0);
			return t->getTransport();
		}
	}
}

MamaTransport* MdsOmTransports::getTransport()
{
	if (transportList.size() == 0) {
		mama_log(MAMA_LOG_LEVEL_ERROR, "MdsOmTransports::getTransport(): error %s list is empty", env->getName());
		throw MdsOmStatus(MDS_OM_STATUS_NO_TRANSPORTS_AVAILABLE);
	}

	MdsOmTransport* t = transportList.at(listIndex);
	if (++listIndex >= transportList.size()) {
		listIndex = 0;
	}
	if (t == NULL) {
		return transportList.at(0)->getTransport();
	} else {
		return t->getTransport();
	}
}

MdsOmTransport* MdsOmTransports::getMdsOmTransport(MamaTransport* transport)
{
	if (transport == NULL) return NULL;

	if (transportList.size() > 0) {
		MdsOmTransportCollection::iterator it = transportList.begin();
		while (it != transportList.end()) {
			MdsOmTransport* t = *it++;
			if (t->getTransport() == transport) return t;
		}
	} else {
		MdsOmTransportMap::const_iterator it = transportMap.begin();
		while (it != transportMap.end()) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				if (t->getTransport() == transport) {
					return t;
				}
			}
			it++;
		}
	}
	return NULL;
}

void MdsOmTransports::getTransports(vector<MamaTransport*>& tlist) const
{
	if (transportList.size() == 0 && transportMap.size() == 0) {
		return;
	}

	if (!isMap()) {
		for (size_t i = 0; i < transportList.size(); i++) {
			tlist.push_back(transportList.at(i)->getTransport());
		}
	} else {
		MdsOmTransportMap::const_iterator it = transportMap.begin();
		while (it != transportMap.end()) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				tlist.push_back(t->getTransport());
			}
			it++;
		}
	}
}

/**
 * Are all transports in connected state?
 */
bool MdsOmTransports::isConnected() const
{
	bool ok = true;
	if (!isMap()) {
		for (size_t i = 0; i < transportList.size(); i++) {
			MdsOmTransport* t = transportList.at(i);
			if (t->isConnected() == false) ok = false;
		}
	} else {
		MdsOmTransportMap::const_iterator it = transportMap.begin();
		while (it != transportMap.end()) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				if (t->isConnected() == false) ok = false;
			}
			it++;
		}
	}
	return ok;
}

/**
 * Is any transport waiting?
 */
bool MdsOmTransports::isConnectWait() const
{
	bool waiting = false;
	if (!isMap()) {
		for (size_t i = 0; i < transportList.size(); i++) {
			MdsOmTransport* t = transportList.at(i);
			if (t->isConnectWait()) waiting = true;
		}
	} else {
		MdsOmTransportMap::const_iterator it = transportMap.begin();
		while (it != transportMap.end()) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				if (t->isConnectWait()) waiting = true;
			}
			it++;
		}
	}
	return waiting;
}

/**
 * Is any transport failed?
 */
bool MdsOmTransports::isConnectFail() const
{
	bool fail = false;
	if (!isMap()) {
		for (size_t i = 0; i < transportList.size(); i++) {
			MdsOmTransport* t = transportList.at(i);
			if (t->isConnectFail()) fail = true;
		}
	} else {
		MdsOmTransportMap::const_iterator it = transportMap.begin();
		while (it != transportMap.end()) {
			MdsOmTransportSet* s = it->second;
			MdsOmTransportCollection::const_iterator itx;
			for (itx = s->transportList.begin(); itx != s->transportList.end(); itx++) {
				MdsOmTransport* t = *itx;
				if (t->isConnectFail()) fail = true;
			}
			it++;
		}
	}
	return fail;
}

}

File: Java/MdsOmHelperClasses/src/com/local/mds/mercury/helper/MdsOmTransports.java
package com.jpmorgan.mds.mercury.helper;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.List;
import java.util.Map;

import com.jpmorgan.mds.mercury.helper.MdsOmTransport.MdsOmTransportStatus;
import com.wombat.mama.Mama;
import com.wombat.mama.MamaLogLevel;
import com.wombat.mama.MamaTransport;

public class MdsOmTransports {
	
	private MdsOmEnv env = null;
	private List<MdsOmTransport> transportList = new ArrayList<MdsOmTransport>();
	private Map<String, MdsOmTransport> transportMap = new Hashtable<String, MdsOmTransport>();
	int listIndex = 0;
	
	public MdsOmTransports() {
	}
	
	public void close() throws Exception {	
		try {
			Mama.log(MamaLogLevel.NORMAL, "MdsOmTransports::close: " + env.getName() + " delete transports map count=" + transportMap.size());
			transportMap.clear();

			Mama.log(MamaLogLevel.NORMAL, "MdsOmTransports::close: " + env.getName() + " delete transports list count=" + transportList.size());
			transportList.clear();

			Mama.log(MamaLogLevel.NORMAL, "MdsOmTransports::close: " + env.getName());
		} catch (Exception status) {
			Mama.log(MamaLogLevel.ERROR, "MdsOmTransports::close: error " + env.getName() + " " + status.toString());
			// Send original exception reference back to caller
			throw status;
		}
	}
	
	public void setEnv(MdsOmEnv env) { this.env = env; }

	public boolean isMap() { return transportMap.size() > 0; }

	public int mapSize() { return transportMap.size(); }

	public int listSize() { return transportList.size(); }

	public void addTransport(MamaTransport transport) {
		// Add to list
		transportList.add(new MdsOmTransport(transport));
	}

	public void addTransport(MamaTransport transport, String source) {
		// Add to map
		transportMap.put(source, new MdsOmTransport(transport));
	}

	public void setStatus(MamaTransport transport, MdsOmTransportStatus status)
	{
		if (isMap()) {
			for (MdsOmTransport t : transportMap.values()) {
				if (t.getTransport() == transport) {
					t.setStatus(status);
					return;
				}
			}
		} else {
			for (MdsOmTransport t : transportList) {
				if (t.getTransport() == transport) {
					t.setStatus(status);
					return;
				}
			}
		}
		Mama.log(MamaLogLevel.ERROR,"MdsOmTransports::setStatus: did not find transport " + transport.getName());
		printTransports();
	}

	public void printTransports() {
		Mama.log(MamaLogLevel.NORMAL, "MdsOmTransports::printTransports: " + env.getName() + " map=" + transportMap.size() + " list=" + transportList.size());
			
		if (isMap()) {
			for (MdsOmTransport t : transportMap.values()) {
				Mama.log(MamaLogLevel.NORMAL, "MdsOmTransports::printTransports: " + t.getTransport().getName());
			}
		} else {
			for (MdsOmTransport t : transportList) {
				Mama.log(MamaLogLevel.NORMAL, "MdsOmTransports::printTransports: " + t.getTransport().getName());
			}
		}
	}

	public MamaTransport getTransport(String source) throws MdsOmStatus {
		if (transportMap.size() == 0 && transportList.size() == 0) {
			Mama.log(MamaLogLevel.ERROR, "MdsOmTransports::getTransport(source): error " + env.getName() + " both collections are empty");
			throw new MdsOmStatus(MdsOmStatus.NO_TRANSPORTS_AVAILABLE);
		}

		if (transportList.size() > 0) {
			return getTransport();
		} else {
			MdsOmTransport t = transportMap.get(source);
			if (t != null) return t.getTransport();
			for (MdsOmTransport t1 : transportMap.values()) {
				return t1.getTransport();
			}
		}
		return null;
	}

	public MamaTransport getTransport() throws MdsOmStatus {
		if (transportList.size() == 0) {
			Mama.log(MamaLogLevel.ERROR, "MdsOmTransports::getTransport(): error " + env.getName() + " list is empty");
			throw new MdsOmStatus(MdsOmStatus.NO_TRANSPORTS_AVAILABLE);
		}

		MdsOmTransport t = transportList.get(listIndex);
		if (++listIndex >= transportList.size()) {
			listIndex = 0;
		}
		if (t == null) {
			return transportList.get(0).getTransport();
		} else {
			return t.getTransport();
		}
	}

	public void getTransports(List<MamaTransport> tlist) {
		if (transportList.size() == 0 && transportMap.size() == 0) {
			return;
		}

		if (!isMap()) {
			for (MdsOmTransport t : transportList) {
				tlist.add(t.getTransport());
			}
		} else {
			for (MdsOmTransport t : transportMap.values()) {
				tlist.add(t.getTransport());
			}
		}
	}


	public boolean isConnected() {
		boolean ok = true;
		if (!isMap()) {
			for (MdsOmTransport t : transportList) {
				if (t.isConnected() == false) ok = false;
			}
		} else {
			for (MdsOmTransport t : transportMap.values()) {
				if (t.isConnected() == false) ok = false;
			}
		}
		return ok;
	}


	public boolean isConnectWait() {
		boolean waiting = false;
		if (!isMap()) {
			for (MdsOmTransport t : transportList) {
				if (t.isConnectWait() == true) waiting = true;
			}
		} else {
			for (MdsOmTransport t : transportMap.values()) {
				if (t.isConnectWait() == true) waiting = true;
			}
		}
		return waiting;
	}


	public boolean isConnectFail() {
		boolean fail = false;
		if (!isMap()) {
			for (MdsOmTransport t : transportList) {
				if (t.isConnectFail() == true) fail = true;
			}
		} else {
			for (MdsOmTransport t : transportMap.values()) {
				if (t.isConnectFail() == true) fail = true;
			}
		}
		return fail;
	}
}
